spectec/src/backend-prose/render.ml

```diff
@@ -304,36 +304,47 @@ and render_iters env iters = List.map (render_iter env) iters |> List.fold_left
 
 (* Expressions and Paths *)
 
-(* fallback when expr cannot be embedded in a math block *)
+(* Invariant: All AL expressions fall into one of the three categories:
+  1. EL-like expression, only containing EL subexpressions
+  2. AL-only expression, possibly containing EL subexpressions 
+  3. pseudo-EL-like expression, containing at least one AL-only subexpression *)
+
+(* Category 1 is translated to EL then rendered by the Latex backend *)
+
+and render_expr env expr = match al_to_el_exp expr with
+  | Some exp -> 
+      (* embedded math blocks cannot have line-breaks *)
+      let newline = Str.regexp "\n" in
+      let sexp = Backend_latex.Render.render_exp env.render_latex exp in
+      let sexp = Str.global_replace newline "" sexp in
+      render_math sexp
+  | None -> render_expr' env expr
+
+(* Categories 2 and 3 are rendered by the prose backend,
+   yet EL subexpressions are still rendered by the Latex backend *)
 
 and render_expr' env expr =
   match expr.it with
-  | Al.Ast.VarE n | Al.Ast.SubE (n, _) -> n
-  | Al.Ast.NumE i -> Int64.to_string i
   | Al.Ast.BoolE b -> string_of_bool b
   | Al.Ast.UnE (NotOp, { it = Al.Ast.IsCaseOfE (e, kwd); _ }) ->
       let se = render_expr env e in
       let skwd = render_math (render_kwd env kwd) in
       sprintf "%s is not of the case %s" se skwd
   | Al.Ast.UnE (NotOp, { it = Al.Ast.IsDefinedE e; _ }) ->
       let se = render_expr env e in
-      sprintf "%s is not defined" se 
+      sprintf "%s is not defined" se
   | Al.Ast.UnE (NotOp, { it = Al.Ast.IsValidE e; _ }) ->
       let se = render_expr env e in
-      sprintf "%s is not valid" se 
-  | Al.Ast.UnE (op, e) ->
+      sprintf "%s is not valid" se
+  | Al.Ast.UnE (op, e) when Option.is_none (al_to_el_unop op) ->
```

Comment:
I instead enforced `render_al_unop` and `render_al_binop` to print out textual (or, not-formula) formats *only*. For instance, `render_al_binop` renders `Al.Ast.MulOp` as "multiplied by" rather than "\\cdot".

So we have a guarantee that inside `render_expr'` (which being their *only* callsites), they will *never* print formula as plaintext.

Ditto for `render_path` mentioned above.